Volume Scheduling Limits

bertinatto opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This PR implements the changes proposed by the KEP [Volume Scheduling Limits](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190408-volume-scheduling-limits.md). 

**Special notes for your reviewer**:

Commits are grouped to make review easier, will squash them later.

**Does this PR introduce a user-facing change?**:

```release-note
Integrated volume limits for in-tree and CSI volumes into one scheduler predicate.
```
CC @msau42, @saad-ali, @davidz627, @jsafrane, @gnufied

gnufied commented on pkg/apis/storage/types.go:
```diff
@@ -350,6 +353,13 @@ type CSINodeDriver struct {
+	// If the field is nil, the node can handle any amount of volumes provided by the driver.
+	Count *int32
+}
+
```
Do we need some validations to ensure negative values can't be set?

gnufied commented on pkg/apis/storage/types.go:
```diff
@@ -335,6 +335,9 @@ type CSINodeDriver struct {
 	NodeID string
 
+	// allocatable represents the resources of a node that are available for scheduling.
+	Allocatable VolumeLimits
```
Since `Allocatable` is a optional field (for now), I think this should be made a pointer?

cwdsuzhou replied:
Seems  pointer is better

gnufied commented on staging/src/k8s.io/api/storage/v1beta1/types.go:
```diff
@@ -337,6 +337,9 @@ type CSINodeDriver struct {
 	NodeID string `json:"nodeID" protobuf:"bytes,2,opt,name=nodeID"`
 
+	// allocatable represents the resources of a node that are available for scheduling.
+	Allocatable VolumeLimits `json:"allocatable" protobuf:"bytes,5,opt,name=allocatable"`
```
missing `+optional` comment

gnufied commented on staging/src/k8s.io/api/storage/v1beta1/types.go:
```diff
@@ -349,7 +352,14 @@ type CSINodeDriver struct {
+type VolumeLimits struct {
+	// Maximum number of volumes provided by the CSI driver that the node can handle.
+	// If the field is nil, the node can handle any amount of volumes provided by the driver.
+	Count *int32 `json:"count,omitempty" protobuf:"varint,1,opt,name=count"`
```
missing `+optional` comment.

gnufied commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -548,10 +547,25 @@ func (nim *nodeInfoManager) installDriverToCSINode(
+		NodeID: driverNodeID,
+		Allocatable: storagev1beta1.VolumeLimits{
+			Count: maxLimit,
+		},
```
We should set this only when attach limit feature gate is enabled.

bertinatto replied:
Fixed

gnufied commented:
/label api-review

liggitt commented:
/assign @thockin

msau42 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager_test.go:
```diff
@@ -431,15 +449,63 @@ func TestInstallCSIDriver(t *testing.T) {
+			name:             "new node with overflown max limit",
+			driverName:       "com.example.csi.driver1",
+			existingNode:     generateNode(nil /*nodeIDs*/, nil /*labels*/, nil /*capacity*/),
+			inputVolumeLimit: math.MaxInt32 + 1,
```
also add a test case == max

msau42 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -555,6 +554,17 @@ func (nim *nodeInfoManager) installDriverToCSINode(
+				Count: &m,
+			}
+		} else {
+			klog.Errorf("Invalid attach limit value %d cannot be added to CSINode object for %q", maxAttachLimit, driverName)
```
It's technically valid from the CSI spec.  Maybe either truncate the limit to the MaxInt32 or keep it as no limit and change the log to a warning with "Exceeded max supported attach limit value..."

msau42 commented on test/e2e/storage/csi_mock_volume.go:
```diff
@@ -353,33 +353,15 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
 
-			Expect(nodeAttachLimit).To(Equal(2))
-
-			_, _, pod1 := createPod()
```
Why remove the pod creations? This will actually test that the scheduler enforces the limit.

gnufied replied:
I think these tests will not pass until we modify the scheduler.

jsafrane replied:
Commenting out + TODO comment would be better in this case.

msau42 commented:
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/storage/csinode/strategy.go should also be modified to drop fields based on feature gates

bertinatto commented on pkg/apis/storage/validation/validation_test.go:
```diff
@@ -1207,6 +1235,20 @@ func TestCSINodeValidation(t *testing.T) {
 			},
 		},
+		{
+			// Volume limits with negative number
```
Done

bertinatto commented:
/test pull-kubernetes-e2e-gce-storage-slow

bertinatto commented:
> https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/storage/csinode/strategy.go should also be modified to drop fields based on feature gates

Done

jsafrane commented on pkg/apis/storage/types.go:
```diff
@@ -350,6 +354,14 @@ type CSINodeDriver struct {
 
+// VolumeLimits is a set of resource limits for scheduling of volumes.
+type VolumeLimits struct {
+	// Maximum number of volumes provided by the CSI driver that the node can handle.
```
I don't like introducing 'attach' term anywhere in this limit. Non-attachable volumes may have limits too.

msau42 replied:
How can we describe this limit in more detail?  Number of unique volumes that can be used by all pods on a node?

jsafrane replied:
"Maximum number of volumes that the driver supports to be used on a node"?

bertinatto replied:
Updated. PTAL

jsafrane commented on pkg/apis/storage/types.go:
```diff
@@ -350,6 +354,14 @@ type CSINodeDriver struct {
 }
 
+// VolumeLimits is a set of resource limits for scheduling of volumes.
+type VolumeLimits struct {
```
`VolumeNodeResources` sounds usable, however I don't like 'AttachLimit'. Non-attachable volumes may have limits too.

gnufied replied:
The link to discussion in KEP - https://github.com/kubernetes/enhancements/pull/942#discussion_r275566976 and https://github.com/kubernetes/enhancements/pull/942#discussion_r273104306

gnufied replied:
We can just call it `Volumes` ? Since key is called `Allocatable`. `Count` will return number of allocatable volumes on the node and `Capacity` could return total size of allocatable volumes on the node?

msau42 replied:
The type as `Volumes` doesn't fit well in the context of the entire `storage.k8s.io` api group.

gnufied commented on pkg/apis/storage/validation/validation.go:
```diff
@@ -329,12 +329,25 @@ func validateCSINodeDriverNodeID(nodeID string, fldPath *field.Path) field.Error
+		return allErrs
+	}
+
+	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(*a.Count), fldPath.Child("count"))...)
```
yeah it was discussed in the KEP. @bsalamat expected this to be int32 since that should be enough to store volume limits.

msau42 replied:
Do we want to validate maxint32 here too then?

bertinatto replied:
Since `*a.Count` is an  `int32`, I think we don't need to check if it's `> MaxInt32`, we only need to check if it's negative.

The worst thing that can happen is someone assigning `Count` with a truncated value (this will result in `Count == 1`, and we'd have the same problem if `Count` was an `int64`):

```
x := int64((math.MaxInt32 + 1) + (math.MaxInt32 + 1) + 1)
allocatable.Count = int32(x)
```

gnufied commented on pkg/apis/storage/types.go:
```diff
@@ -350,6 +354,14 @@ type CSINodeDriver struct {
+	// Maximum number of volumes provided by the CSI driver that the node can handle.
+	// If the field is nil, the node can handle any amount of volumes provided by the driver.
+	// +optional
+	Count *int32
```
link to discussion about naming it `Count` - https://github.com/kubernetes/enhancements/pull/942#discussion_r279118467

msau42 replied:
Maybe Count is fine as long as we can comment exactly what it means.

Just on its own it's too vague.  Is it number of mounts? number of attaches? what if multiple pods on the same node share the volume?  What is and what is not counted?

bertinatto replied:
(Resolving this in favour of thread above)

bertinatto replied:
Updated. PTAL

bertinatto commented:
/test pull-kubernetes-integration

msau42 commented on pkg/registry/storage/csinode/strategy.go:
```diff
@@ -54,16 +62,21 @@ func (csiNodeStrategy) Validate(ctx context.Context, obj runtime.Object) field.E
+	csiNode := obj.(*storage.CSINode)
+	if !utilfeature.DefaultFeatureGate.Enabled(features.AttachVolumeLimit) {
+		for _, csiDriver := range csiNode.Spec.Drivers {
+			csiDriver.Allocatable = nil
```
The guidance we've been following for all other feature-gated fields, is that if it's already set, we continue to let it be set.  This handles cases where the feature was on, and then the cluster was downgraded to a version where the feature is off.

bertinatto replied:
SGTM. Removed

msau42 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -555,6 +554,19 @@ func (nim *nodeInfoManager) installDriverToCSINode(
+			m := int32(maxAttachLimit)
+			driverSpec.Allocatable = &storagev1beta1.VolumeLimits{Count: &m}
+		} else {
+			klog.Errorf("Invalid attach limit value %d cannot be added to CSINode object for %q", maxAttachLimit, driverName)
```
Will these log messages be repeated frequently? If so, maybe we should have a higher log level.  Or should we actually return an error? Will that cause driver registration to fail?

bertinatto replied:
I don't expect them to repeat frequently, only during the registration of a driver that has an invalid max volumes number.

> Or should we actually return an error? Will that cause driver registration to fail?

Yeah, the registration would fail...

msau42 replied:
Should we fail registration?  The driver is passing us an invalid value.

bertinatto commented:
/test pull-kubernetes-e2e-gce-100-performance

msau42 commented on pkg/apis/storage/types.go:
```diff
@@ -350,6 +354,16 @@ type CSINodeDriver struct {
+	// Maximum number of unique volumes provisioned by the CSI driver that can be used on a node.
+	// A volume that is both attached and mounted on a node is considered to be used once, not twice.
+	// The same rule applies for an unique volume that is shared among multiple pods on the same node.
+	// This field can be nil if there is no limit of volumes supported on a node.
```
How about:
If this field is nil, then the supported number of volumes on this node is unbounded

bertinatto replied:
Changed

msau42 commented on pkg/apis/storage/validation/validation_test.go:
```diff
@@ -991,6 +991,34 @@ func TestCSINodeValidation(t *testing.T) {
+			},
+		},
+		{
+			// Volume limits with positive number
```
One more test case for 0

msau42 commented on pkg/registry/storage/csinode/strategy.go:
```diff
@@ -54,15 +62,13 @@ func (csiNodeStrategy) Validate(ctx context.Context, obj runtime.Object) field.E
 }
 
-// PrepareForUpdate sets the Status fields which is not allowed to be set by an end user updating a CSINode
 func (csiNodeStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
```
Sorry I meant that the logic here for dropping the allocatable field should check for feature gate and if the field is already set.  You can use the [storageclass prepareforupdate](https://github.com/kubernetes/kubernetes/blob/d75da7e17e234277f072f3e1324d8a475343e26a/pkg/registry/storage/storageclass/strategy.go#L66) as an example.

bertinatto replied:
Done. I used a map to avoid repeatedly iterating over the drivers of the old object.

apelisse commented on pkg/apis/storage/types.go:
```diff
@@ -350,6 +354,16 @@ type CSINodeDriver struct {
+	// The same rule applies for an unique volume that is shared among multiple pods on the same node.
+	// This field can be nil if there is no limit of volumes supported on a node.
+	// +optional
+	Count *int32
```
Sorry, more bikeshedding on naming ... What about `MaxCount`?

apelisse commented on pkg/apis/storage/validation/validation_test.go:
```diff
@@ -1360,17 +1418,39 @@ func TestCSINodeUpdateValidation(t *testing.T) {
 			},
 		},
+		{
+			// invalid change max volumes limit
```
That's a good name for the variable `MaxVolumesLimit` :-)

apelisse commented on staging/src/k8s.io/api/storage/v1beta1/types.go:
```diff
@@ -349,7 +353,17 @@ type CSINodeDriver struct {
 	// This can be empty if driver does not support topology.
 	// +optional
-	TopologyKeys []string `json:"topologyKeys" protobuf:"bytes,3,rep,name=topologyKeys"`
+	TopologyKeys []string `json:"topologyKeys" protobuf:"bytes,4,rep,name=topologyKeys"`
```
I don't think you can change protobuf field numbers like this, that's going to break existing serialized objects. (use 4 for allocatable)

thockin commented:
API Approved

/approve

bertinatto commented on pkg/apis/storage/types.go:
```diff
@@ -350,6 +354,16 @@ type CSINodeDriver struct {
 
+// VolumeNodeResources is a set of resource limits for scheduling of volumes.
+type VolumeNodeResources struct {
+	// Maximum number of unique volumes provisioned by the CSI driver that can be used on a node.
```
Fixed

msau42 commented:
/hold
wait until comments are addressed and scheduling changes are in too

msau42 commented on pkg/apis/storage/validation/validation_test.go:
```diff
@@ -1229,6 +1271,10 @@ func TestCSINodeValidation(t *testing.T) {
+	return &i
+}
+
 func TestCSINodeUpdateValidation(t *testing.T) {
```
Instead of modifying existing test cases to add allocatable, can we have new test cases for allocatable?

Toggle feature on and off:
- Before no allocatable, try to set allocatable
- Before allocatable, update same allocatable
- Before allocatable, try to change allocatable
- Before allocatable, try to remove allocatable

bertinatto replied:
Yeah, I missed a few tests here. Added

What's the difference between `update same (some?) allocatable` and `try to change allocatable`?

bertinatto commented on pkg/registry/storage/csinode/strategy.go:
```diff
@@ -54,15 +62,13 @@ func (csiNodeStrategy) Validate(ctx context.Context, obj runtime.Object) field.E
 	return errs
 }
 
-// Canonicalize normalizes the object after validation.
```
I think I fat-fingered this one. Added the comment back

bertinatto commented on pkg/apis/storage/validation/validation_test.go:
```diff
@@ -1207,6 +1235,20 @@ func TestCSINodeValidation(t *testing.T) {
+					{
+						Name:         "io.kubernetes.storage.csi.driver",
+						NodeID:       nodeID,
+						Allocatable:  &storage.VolumeNodeResources{Count: makeInt32Ptr(-1)},
```
Changed to use `utilpointer.Int32Ptr()` instead.

bertinatto commented:
/retest

bertinatto commented:
/skip

bertinatto commented:
/retest

bertinatto commented:
/skip

bertinatto commented on pkg/apis/storage/validation/validation_test.go:
```diff
@@ -1360,13 +1409,90 @@ func TestCSINodeUpdateValidation(t *testing.T) {
 					{
 						Name:         "io.kubernetes.storage.csi.driver-1",
-						NodeID:       "nodeB",
+						NodeID:       nodeID,
```
This case was meant to test topology keys, not node id.

bertinatto commented:
/test pull-kubernetes-verify
```
openapi-spec:

/home/prow/go/src/k8s.io/kubernetes/api/openapi-spec is out of date. Please run hack/update-openapi-spec.sh
```

The openapi-spec is updated and verify doesn't fail for me. I don't know why this test keeps failing.

gnufied commented:
I think you do need to run `/hack/update-openapi-spec.sh` ? When I do that it does create a diff:


```
iff --git a/api/openapi-spec/swagger.json b/api/openapi-spec/swagger.json
index f1c5db1655..18261d0c21 100644
--- a/api/openapi-spec/swagger.json
+++ b/api/openapi-spec/swagger.json
@@ -18421,7 +18421,7 @@
   },
   "info": {
     "title": "Kubernetes",
-    "version": "v1.15.0"
+    "version": "v1.16.0"
   },
   "paths": {
     "/api/": {
```

bertinatto committed:
Add volume limits API changes

bertinatto committed:
Auto-generated code for volume limits API

bertinatto committed:
Update nodeinfomanager to store volume limits in CSINode

msau42 commented on pkg/registry/storage/csinode/strategy_test.go:
```diff
@@ -36,6 +37,7 @@ func getValidCSINode(name string) *storage.CSINode {
 					Name:         "valid-driver-name",
 					NodeID:       "valid-node",
 					TopologyKeys: []string{"company.com/zone1", "company.com/zone2"},
+					Allocatable:  &storage.VolumeNodeResources{Count: utilpointer.Int32Ptr(10)},
```
a few more test cases for prepareForUpdate:

Feature gate off:
* allowed - Old object: allocatable not set; New object: allocable not set
* dropped - Old object: allocatable not set; New object: allocable set

Feature gate on:
* allowed - Old object: allocatable not set; New object: allocable not set
* allowed - Old object: allocatable not set; New object: allocable set

bertinatto replied:
Added

msau42 commented on staging/src/k8s.io/api/storage/v1beta1/types.go:
```diff
@@ -350,6 +350,21 @@ type CSINodeDriver struct {
+	// A volume that is both attached and mounted on a node is considered to be used once, not twice.
+	// The same rule applies for an unique volume that is shared among multiple pods on the same node.
+	// This field can be nil if there is no limit of volumes supported on a node.
+	// If this field is nil, then the supported number of volumes on this node is unbounded.
```
Make sure to sync the comments from all the types.go files

bertinatto replied:
Fixed

msau42 reviewed (commented):
api changes generally lgtm, just some minor comments.

waiting for scheduler changes now.

msau42 commented:
/assign @jsafrane @msau42

bertinatto committed:
Update scheduler to use volume limits from CSINode

bertinatto commented:
/test pull-kubernetes-e2e-gce

bertinatto commented:
/test pull-kubernetes-e2e-gce-csi-serial

bertinatto commented:
/test pull-kubernetes-e2e-gce-100-performance

bertinatto commented:
/test pull-kubernetes-integration

bertinatto commented:
/skip

bertinatto commented:
/test pull-kubernetes-integration

bertinatto commented:
/test pull-kubernetes-integration

jsafrane commented on pkg/scheduler/nodeinfo/node_info.go:
```diff
@@ -286,6 +287,14 @@ func (n *NodeInfo) Node() *v1.Node {
 }
 
+// CSINode returns overall CSI-related information about this node.
+func (n *NodeInfo) CSINode() *storagev1beta1.CSINode {
```
This function is not used.

jsafrane commented on pkg/scheduler/nodeinfo/node_info.go:
```diff
@@ -473,6 +482,17 @@ func (n *NodeInfo) Clone() *NodeInfo {
 func (n *NodeInfo) VolumeLimits() map[v1.ResourceName]int64 {
 	volumeLimits := map[v1.ResourceName]int64{}
+
+	if n.csiNode != nil {
```
IMO, if CSINode exists and it contains the driver, we should use it and not to fall back to `node.status.allocatable`. Should be the logic reversed, collect `node.allocatable` first and overwrite it with `csiNode.*.allocatable` later?

bertinatto commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -385,6 +449,14 @@ func AddAllEventHandlers(
 		},
 	)
 
+	csiNodeInformer.Informer().AddEventHandler(
```
Added

bertinatto commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -57,6 +56,8 @@ var (
 		Factor:   5.0,
 		Jitter:   0.1,
 	}
+	attachVolumeLimitEnabled = utilfeature.DefaultFeatureGate.Enabled(features.AttachVolumeLimit)
```
Added back

bertinatto commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -113,21 +114,17 @@ func (nim *nodeInfoManager) InstallCSIDriver(driverName string, driverNodeID str
 		nodeUpdateFuncs = append(nodeUpdateFuncs, updateTopologyLabels(topology))
 	}
 
-	if utilfeature.DefaultFeatureGate.Enabled(features.AttachVolumeLimit) {
```
According to the [KEP](https://github.com/kubernetes/enhancements/blob/93a90a4f5d255eb60d5c80cca30fc20d13e0ef5d/keps/sig-storage/20190408-volume-scheduling-limits.md#design-details), we should keep using this feature gate for this functionality:

```
Existing feature gate AttachVolumeLimit will be re-used for implementation of this KEP. The feature is already beta and is enabled by default.
```
In fact, the predicate still uses it: 

https://github.com/kubernetes/kubernetes/blob/4db3fccd6f1c5ae8deb687985a2c8be76e4d4ab5/pkg/scheduler/algorithm/predicates/csi_volume_predicate.go#L54-L57

bertinatto replied:
The KEP also [mentions](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190408-volume-scheduling-limits.md#implementation-detail-for-all-csi-drivers) that the Kubelet should stop adding the limits to the `Node` object:

```
Scheduler will respect Node.status.allocatable and Node.status.capacity for CSI volumes if CSINode object is not available or has missing entry in CSINode.spec.drivers[xyz].allocatable during a deprecation period but kubelet will stop populating Node.status.allocatable and Node.status.capacity for CSI volumes. 
```

I have the impression that the KEP's idea was to do something similar to what we've done with the `CSINodeInfo` object/gate: we migrated it to beta and changed the underlying implementation. `AttachVolumeLimits` is already beta though, so I don't know if we can do the same here.

Thoughts?

msau42 replied:
I think having kubelet stop reporting limits through Node object is fine as long as the scheduler still supports both for up to 2 version skew.

bertinatto replied:
> +1 I'm not sure if Node object gets deleted/recreated during node upgrade, but may be good to keep cleanup logic just in case.

Added the cleanup back

bertinatto replied:
To summarize: kubelet stop reporting limits through Node object, but cleanup logic is still there.

bertinatto commented:
/test pull-kubernetes-bazel-test

bertinatto commented:
/test pull-kubernetes-e2e-gce

bertinatto commented:
/test pull-kubernetes-e2e-gce

gnufied commented:
pinging scheduler folks for review


/assign @ravisantoshgudimetla @bsalamat

ravisantoshgudimetla commented on pkg/scheduler/factory/factory.go:
```diff
@@ -232,7 +236,8 @@ type ConfigFactoryArgs struct {
 	PdbInformer                    policyinformers.PodDisruptionBudgetInformer
-	StorageClassInformer           storageinformers.StorageClassInformer
+	StorageClassInformer           storageinformersv1.StorageClassInformer
+	CSINodeInformer                storageinformersv1beta1.CSINodeInformer
```
I am not 100% sure about the performance of scheduler with this change and since this feature is beta, this will be enabled by default. So, I feel less confident about this change going into 1.15.

jsafrane replied:
There were no scheduler benchmarks with volumes, I am preparing one at https://github.com/kubernetes/kubernetes/pull/77397.

Current numbers (5000 nodes each with CSINode instance, 1000 pods each with a CSI volume, with all alpha gates disabled, two measurements each):

With this PR:    30583423 and 37684721 ns/op
Without this PR: 33340480 and 36254432 ns/op

These numbers are jumping like crazy... will do proper stats on a stable system tomorrow.

jsafrane replied:
Finally I got some real numbers:

| Volume             | Original ns/op | With this PR ns/op | Increase |
| ------------------ | -------------- | ------------------ | -------- |
| In-tree volume     |  77432515      |  78835497          | 101.81%  |
| CSI volume         |  64654527      |  70410337          | 108.90%  |
| Migrated volume    |  80161751      |  78255729          | 97.62%   |

All numbers are form tests BenchmarkScheduling*PVs/5000Nodes/1000Pods-4 on 4-core GCE machine, 10 samples each (variation is low, +/- 2-3%)

CSI scheduling is 8% slower, the rest is IMO fine.

jsafrane replied:
I ran the tests with current version + enabled benchmark of a pod with Secrets volume (in two flavours, one with no CSINode instances and second with CSINode for each Node), results are not so nice:

Test | Before (ns/op) | After (ns/op) | Increase
-- | -- | -- | --
BenchmarkSchedulingInTreePVs | 79628055.7 | 79433974.9 | -0.24%
BenchmarkSchedulingCSIPVs | 65328092.6 | 70548380.8 | 7.99%
BenchmarkSchedulingMigratedInTreePVs | 80408231.1 | 80375741.1 | -0.04%
BenchmarkSchedulingSecrets | 30153507.9 | 33890242.5 | 12.39%
BenchmarkSchedulingSecretsWithCSINodes | 30181808.4 | 36221780.5 | 20.01%

Main culprits:

- feature gate checks
- VolumeLimits() call

We're working on optimizations.

msau42 replied:
BenchmarkSchedulingSecretsWithCSINodes mistakenly turned on migration alpha gate while the others had it disabled.

I'm testing out some optimizations.

liggitt commented:
fyi, ran this against the [in-progress API compatibility tests](https://github.com/kubernetes/kubernetes/pull/78309) and the changes in this PR do not affect round-tripping of 1.14.x-level data, so 👍

gnufied commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -134,33 +137,33 @@ func (c *CSIMaxVolumeLimitChecker) filterAttachableVolumes(
+		volumeUniqueName := fmt.Sprintf("%s/%s", driverName, volumeHandle)
 		volumeLimitKey := volumeutil.GetCSIAttachLimitKey(driverName)
-		result[volumeHandle] = volumeLimitKey
+		result[volumeUniqueName] = volumeLimitKey
```
So this code will not prevent scheduling of pods to a node which does not have CSI driver installed yet. The KEP states that - for migrated plugins we should immediately stop scheduling of pods to nodes where driver is not installed and for other (non-migrated drivers) after 2 releases we are going to stop scheduling of pods to the node where driver is not installed.

gnufied replied:
I am probably missing something but where did you add that check? AFAICT - we are still not preventing pod scheduling if migration is enabled for a CSI volume and driver is not installed on the node.

gnufied commented on pkg/scheduler/algorithm/predicates/utils.go:
```diff
@@ -87,3 +94,67 @@ func SetPredicatesOrderingDuringTest(value []string) func() {
+// isCSIMigrationOn returns a boolean value indicating whether
+// the CSI migration has been enabled for a particular storage plugin.
+func isCSIMigrationOn(csiNode *storagev1beta1.CSINode, pluginName string) bool {
+	if csiNode == nil || len(pluginName) == 0 {
```
So if driver is not installed on the node, we consider migration to be off on the node? Is that accurate - because this means that pods will still be scheduled on the node and volume operations will be handled by in-tree plugins but I thought that is not what we want to do.

gnufied replied:
cc @davidz627

davidz627 replied:
> So if driver is not installed on the node, we consider migration to be off on the node

This is not accurate.

Determining whether migration is on for a plugin for Attach/Detach involves 2 things:
1) Migration is "on" on the master (feature flags)
2) Migration is "on" on the node which the volume is scheduled to (can be determined through the annotation "storage.alpha.kubernetes.io/migrated-plugins" on the CSINode API object for that node)

For all other operations determining whether migration is on for a plugin involves just:
1) Migration is "on" on the master/node wherever the operation is performed (can just be determined by looking at the feature flags)

Driver installed/not installed does not factor into the determination of migrated or not. If we determine a plugin is migrated with the above criteria, if the driver is not installed generally we throw an error mentioning that the driver was required but not installed.

gnufied replied:
okay thanks for clarifying. I think `isCSIMigrationOn` check is okay. But still we aren't doing anything in the code here to prevent scheduling of pods for which driver is not installed (and migration is on). This should be fixed.

gnufied commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -590,6 +617,13 @@ var CinderVolumeFilter = VolumeFilter{
 		return "", false
 	},
+
+	IsMigrated: func(csiNode *storagev1beta1.CSINode) bool {
```
I do not think this check implements KEP proposal. The KEP says:

>old predicates will be modified to perform an additional check of `CSINode` object at very beginning of their code. If they detect that `CSINode` object contains limit of same underlying volume type or CSI migration has been enabled for the volume, the old predicate will return early(with success) and `MaxCSIVolumeCountPred` will be responsible for counting both CSI and in-tree volumes of same type.


But what this check does is - it only defers to CSI predicate when migration is enabled *AND* `CSINode` object has volume limits for migrated driver. I think originally we wanted to ensure that, volume limits for both in-tree and CSI driver is counted correctly even if both are in-use on a node.

bertinatto replied:
I missed that, and it should be easy to fix but this also implies that the old predicates will defer to the CSI predicate if `CSINode` contains volume limits, even if the migration is disabled.

It means that the limits of an in-tree volume will change if someone installs a CSI driver (that happens to be the "migration counterpart" of the given in-tree volume).

jsafrane replied:
I still don't like moving limits between CSINode and Node back and forth as a driver is installed, I think that the KEP is wrong in this part. See @msau42's comment at https://github.com/kubernetes/enhancements/pull/942#discussion_r277060459, IMO her comment and the code in this PR it is the right thing to do.

Shall we fix the KEP?

gnufied replied:
The reason for doing what we did was - to ensure that we count volumes correctly even if user is using CSI and in-tree volumes simultaneously in the cluster without explicitly enabling in-tree migration. Like imagine a user is using 20 in-tree EBS PVs and 30 CSI EBS PVs on a node, in which case if we kept volume counting to different predicates then we will not have accurate count. It was one of the goals.

Having said that - we can perhaps relax that requirement for now because I am not sure if mixed volumes use without explicitly enabling in-tree migration should be "supported" or not.

msau42 replied:
The current (old) csi predicate already doesn't handle this case. I think we can consider supporting this use case if needed in the future, but I would prefer to keep the logic as simple as possible for now. It's already quite complex with all the cases that we need to handle.

This is only a temporary state while migration is not on yet.

gnufied replied:
Yeah I am fine with dropping this requirement from KEP.

bertinatto replied:
Updated, PTAL.

gnufied commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -169,9 +172,36 @@ func (c *CSIMaxVolumeLimitChecker) getCSIDriver(pvc *v1.PersistentVolumeClaim) (
+			return "", ""
+		}
+
+		csiSource = csiPV.Spec.PersistentVolumeSource.CSI
```
Do we have a scenario where both CSI and in-tree driver can be in-use on a node? If yes - this code will not count them correctly. cc @davidz627 @jsafrane

davidz627 replied:
Do you mean when migration is off and we have the in-tree plugin and its "migration" CSI Driver both be in use on the node?

This is possible.

gnufied commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -385,6 +451,16 @@ func AddAllEventHandlers(
+			},
+		)
+	}
+
```
Thanks for adding this here, but I am not 100% sure if `CSINode` object should be cached inside `NodInfo` object. I will defer to @ravisantoshgudimetla @bsalamat on this.

gnufied commented on pkg/scheduler/nodeinfo/node_info.go:
```diff
@@ -471,11 +482,23 @@ func (n *NodeInfo) Clone() *NodeInfo {
+				k := v1.ResourceName(volumeutil.GetCSIAttachLimitKey(d.Name))
+				volumeLimits[k] = int64(*d.Allocatable.Count)
+			}
+		}
```
This is kinda tricky. This will mean that, we may not be able to distinguish between if driver is installed on the node or volume limit is coming from `Node` object.

davidz627 commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -572,6 +595,10 @@ var AzureDiskVolumeFilter = VolumeFilter{
 	},
+
+	IsMigrated: func(csiNode *storagev1beta1.CSINode) bool {
+		return false
```
Azure Disk/File translation libs just merged

bertinatto replied:
Added

davidz627 commented on pkg/scheduler/algorithm/predicates/utils.go:
```diff
@@ -87,3 +94,67 @@ func SetPredicatesOrderingDuringTest(value []string) func() {
+		if !utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationGCE) {
+			return false
+		}
+	case csilibplugins.CinderInTreePluginName:
```
Is it possible to get the plugin itself and call the `IsMigratedToCSI()` function on the plugin interface? I want to avoid special casing these feature gates checking everywhere

jsafrane replied:
No, scheduler does not import volume plugins. From some reason scheduler people are afraid of all their dependencies (incl. cloud providers and whatnot).

liggitt commented:
FYI, we just added tests to verify we don't break compatibility with previously serialized data in #78309. Part of that keeps up-to-date serializations for current API types in `staging/src/k8s.io/api/testdata/HEAD`. That means that PRs that add fields (like this one), need to update those fixtures.

When this is rebased on master, run the following to update the CSINode fixtures:
```sh
UPDATE_COMPATIBILITY_FIXTURE_DATA=true go test ./vendor/k8s.io/api -run //HEAD
```

bertinatto commented:
> FYI, we just added tests to verify we don't break compatibility with previously serialized data in #78309. Part of that keeps up-to-date serializations for current API types in staging/src/k8s.io/api/testdata/HEAD. That means that PRs that add fields (like this one), need to update those fixtures.

Thanks, updated.

jsafrane commented on pkg/scheduler/nodeinfo/node_info.go:
```diff
@@ -471,11 +482,23 @@ func (n *NodeInfo) Clone() *NodeInfo {
 // VolumeLimits returns volume limits associated with the node
 func (n *NodeInfo) VolumeLimits() map[v1.ResourceName]int64 {
```
As a relatively low-hanging optimization fruit, volume limits for each node can be pre-calculated in informer callbacks (`AddCSINode` / `UpdateCSINode` + their `Node` equivalents). 
Profiler says that `VolumeLimits()` call itself takes 3s out of 8s of whole `attachableLimitPredicate()`.

Furthermore, list of migrated plugins can be pre-parsed for each node there.

I will defer to sig-scheduling to judge if current performance is OK or it needs to be fixed as part of this PR.

msau42 replied:
I'm surprised. The number of objects being iterated over per call should not be very large.  I would have suspected the pod iteration to take the most cycles.

Huang-Wei replied:
@jsafrane Is the perf overhead applied in regular case (without PV/PVC involved)?

jsafrane replied:
> I'm surprised. The number of objects being iterated over per call should not be very large

It is not large, but it is called for every combination of Node and Pod, which is lot of calls. Even simple feature gate check there takes ~0.5s out of this 8 seconds increase.

> Is the perf overhead applied in regular case (without PV/PVC involved)?

Good question! Yes, **all pods with volumes (incl. Secrets) gets such overhead, if CSINode object exists**.

In theory, any volume can be migrated to CSI, so we must know limits reported by the driver. In practice, we won't migrate ephemeral volumes like Secrets, ConfigMap, EmptyDir and so on, some (dirty?) shortcuts are possible. For example, caching of `VolumeLimits` value in `NodeInfo` as I suggested above.

ravisantoshgudimetla replied:
>  Profiler says that VolumeLimits() call itself takes 3s out of 8s of whole attachableLimitPredicate().

To be clear, 8s is avg time taken by `VolumeLimits()` per scheduling of a pod?

msau42 commented on pkg/scheduler/nodeinfo/node_info.go:
```diff
@@ -471,11 +482,23 @@ func (n *NodeInfo) Clone() *NodeInfo {
+		for i := range n.csiNode.Spec.Drivers {
+			d := n.csiNode.Spec.Drivers[i]
+			if d.Allocatable != nil && d.Allocatable.Count != nil {
+				k := v1.ResourceName(volumeutil.GetCSIAttachLimitKey(d.Name))
```
Should we be calling GetCSIAttachLimitKey()? It does the hashing thing when driver name is too big, which the new design should handle.  We should also make sure that non-migration CSI drivers will use the correct resource name.

gnufied replied:
I think this is to ensure that, both new driver names coming from `CSINode` and old names coming from `Node` object are consolidated with unique keys in the `volumeLimits` map. Otherwise, we will end up with 2 values for migrated plugins. 

As a interim step I think it is fine until we can stop recognizing limits coming from node objects..

msau42 replied:
K let's just add a comment to clean this up later

bertinatto replied:
Added TODO

msau42 commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -146,21 +150,19 @@ func (c *CSIMaxVolumeLimitChecker) filterAttachableVolumes(
 }
 
-func (c *CSIMaxVolumeLimitChecker) getCSIDriver(pvc *v1.PersistentVolumeClaim) (string, string) {
+func (c *CSIMaxVolumeLimitChecker) getCSIDriverInfo(csiNode *storagev1beta1.CSINode, pvc *v1.PersistentVolumeClaim) (string, string) {
```
Add a comment describing the function and its return values

msau42 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -140,15 +137,14 @@ func (nim *nodeInfoManager) InstallCSIDriver(driverName string, driverNodeID str
 	}
 
 	err := nim.updateNode(
-		removeMaxAttachLimit(driverName),
```
+1 I'm not sure if Node object gets deleted/recreated during node upgrade, but may be good to keep cleanup logic just in case.

msau42 reviewed (commented):
Scheduling part needs many unit tests.  There's a lot of backwards compatibility scenarios we need to handle.

msau42 commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -143,8 +143,9 @@ func (c *CSIMaxVolumeLimitChecker) filterAttachableVolumes(
 		if driverName == "" || volumeHandle == "" {
 			continue
 		}
+		volumeUniqueName := fmt.Sprintf("%s/%s", driverName, volumeHandle)
```
I'm unsure if this is case we need to handle.   I believe volumeID must be unique, so we may need to fix mock driver if it's not following this

jsafrane replied:
In theory, different CSI drivers may use the same VolumeID for different volumes.

msau42 commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -70,7 +73,7 @@ func (c *CSIMaxVolumeLimitChecker) attachableLimitPredicate(
 	// a map of unique volume name/csi volume handle and volume limit key
 	newVolumes := make(map[string]string)
-	if err := c.filterAttachableVolumes(pod.Spec.Volumes, pod.Namespace, newVolumes); err != nil {
+	if err := c.filterAttachableVolumes(nodeInfo, pod.Spec.Volumes, pod.Namespace, newVolumes); err != nil {
```
We're missing a check (probably somewhere around L108) that limit should == 0 if an entry doesn't exist.

We may be able to make future optimizations around this area too to not have to count volumes if no CSI drivers are installed on the node.

msau42 replied:
Actually there's two cases we need to handle:
- csi driver not installed: limit is 0
- csi driver installed but doesn't report limits: limit is infinite

bertinatto replied:
Can you take a look if https://github.com/kubernetes/kubernetes/pull/77595/commits/e7137eaa9b5795e4e4da6cd9c93014eebf993711 addresses this?

I'm assuming that a `0` limit means no limit. The specs says that the CO should decide how many volumes can be published if the driver reports `.

bertinatto replied:
Sorry, new commit: https://github.com/kubernetes/kubernetes/pull/77595/commits/07651ea160b09b7f9637acb25cb145ce41dfa726

bertinatto replied:
CC @gnufied

gnufied replied:
This should be fixed now.

bertinatto commented:
/test pull-kubernetes-e2e-gce

misterikkit reviewed (commented):
sending a partial review since I have to run to a meeting.

misterikkit commented on pkg/scheduler/algorithm/predicates/utils.go:
```diff
@@ -87,3 +94,71 @@ func SetPredicatesOrderingDuringTest(value []string) func() {
+	if csiNode == nil || len(driverName) == 0 {
+		return false
+	}
+	for i := range csiNode.Spec.Drivers {
```
Why not write this as `for _, d := range csiNode.Spec.Drivers {` ?

misterikkit replied:
Ah, I wasn't thinking about the struct copies. If Drivers was a slice of pointers, then copying would be cheap.

ravisantoshgudimetla replied:
Where are we using this function?

gnufied replied:
We are not using this function anywhere, I will remove it.

misterikkit commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -474,6 +478,11 @@ func (c *MaxPDVolumeCountChecker) predicate(pod *v1.Pod, meta PredicateMetadata,
 		return true, nil, nil
 	}
 
+	// If a plugin has been migrated to a CSI driver, defer to the CSI predicate.
```
Does the CSI predicate already implement per node volume limits? I don't see that update in this PR.

misterikkit reviewed (commented):
I don't fully understand the migration logic here, but I think the part where we cache the CSINode in our NodeInfo struct is the right approach. Predicate evaluation is performed with a snapshot of the NodeInfo so that we have a consistent view during evaluation.

If storage people are happy with the predicate logic, then I think this can be merged.

msau42 commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -169,9 +171,40 @@ func (c *CSIMaxVolumeLimitChecker) getCSIDriver(pvc *v1.PersistentVolumeClaim) (
+			return "", ""
+		}
+
+		csiPV, err := csilib.TranslateInTreePVToCSI(pv)
```
Are we also handling inline (non-pv) migration?

msau42 replied:
That's fine. Migration is still alpha this release. I think most important is to convert existing CSI limits to use the new csinode field.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -19,9 +19,11 @@ package predicates
 	"fmt"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
please remove `v1`

liggitt replied:
This is autoadded by goimports. It's not worth fighting the tool, in my opinion

Huang-Wei replied:
Yes, but it's not worth introducing a dummy alias. I'm now turning to `gofmt`, and I have a long post at https://groups.google.com/d/msg/kubernetes-dev/Y2ShVRrU4xM/LKfaiFqAAwAJ explaining how to configure in vscode.

misterikkit replied:
I agree that it's annoying, but it changes nothing else in the code. I saw it in this PR and decided not to mention it since that's just extra work for the author and us the reviewers.

Incidentally I recall sending feedback on this behavior and getting a "working as intended" response, but I couldn't find that response when I looked for it. I think the tool assumes that our module is actually named "core" and uses semver so that major versions have different import paths. Therefore, the `v1` alias is an indication that no, actually our module is named "v1".

I know it's convention not to alias this import, but I think if we started aliasing it as `core` that things like `core.Pod` would make more sense.

misterikkit replied:
https://github.com/golang/go/issues/30051

bertinatto commented on pkg/scheduler/algorithm/predicates/utils.go:
```diff
@@ -87,3 +94,71 @@ func SetPredicatesOrderingDuringTest(value []string) func() {
+	// The plugin name should be listed in the CSINode object annotation.
+	// This indicates that the plugin has been migrated to a CSI driver in the node.
+	csiNodeAnn := csiNode.GetAnnotations()
+	if csiNodeAnn == nil {
```
I think the idea was to fail fast, i.e., avoiding creating a new empty set and checking if it contains an element.

bertinatto commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -538,6 +547,13 @@ var EBSVolumeFilter = VolumeFilter{
+		if !isCSIMigrationOn(csiNode, csilibplugins.AWSEBSInTreePluginName) {
+			return false
+		}
+		return hasVolumeLimits(csiNode, csilibplugins.AWSEBSDriverName)
```
I needed to wrap `isCSIMigrationOn` into the volume plugin filter because each in-tree volume should be checked against a different CSI driver (if the plugin is migrated).

> Can we remove hasVolumeLimits? If migration is on, then we should always use CSI predicate and not fallback to in-tree predicate regardless of if driver is installed or not.

The KEP states that we should check if the driver is installed or not:

> As described in section above - old predicates will be modified to perform an additional check of CSINode object at very beginning of their code. If they detect that CSINode object contains limit of same underlying volume type or CSI migration has been enabled for the volume, the old predicate will return early(with success) and MaxCSIVolumeCountPred will be responsible for counting both CSI and in-tree volumes of same type.

Should we drop this requirement? CC @gnufied

gnufied replied:
As discussed above - since we are not going to switch counting of volumes from old predicate to new predicate if CSI driver is installed but migration is off, we may drop this check from here.

bertinatto replied:
Dropped

ravisantoshgudimetla commented on pkg/scheduler/factory/factory.go:
```diff
@@ -262,10 +267,17 @@ func NewConfigFactory(args *ConfigFactoryArgs) Configurator {
+
+	// csiNodeInformer is only enabled through AttachVolumeLimit feature gate
+	var csiNodeLister storagelistersv1beta1.CSINodeLister
+	if args.CSINodeInformer != nil {
```
Where are we passing nil for CSINodeInformer, if featuregates are disabled?

misterikkit replied:
+1

bertinatto replied:
I think the comment above is not accurate. Removed.

I'm keeping the check because the informer is an interface and theory could be `nil` (the `StorageClassInformer` above also checks that).

ravisantoshgudimetla commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -61,37 +64,37 @@ func (c *CSIMaxVolumeLimitChecker) attachableLimitPredicate(
-	}
-
 	// a map of unique volume name/csi volume handle and volume limit key
 	newVolumes := make(map[string]string)
```
I think it's better to check if nodeInfo is nil here as well before passing on the pointer.

ravisantoshgudimetla commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -61,37 +64,37 @@ func (c *CSIMaxVolumeLimitChecker) attachableLimitPredicate(
 		return true, nil, nil
 	}
 
+	nodeVolumeLimits := nodeInfo.VolumeLimits()
```
Why can't we have this check earlier and exit earlier?

bertinatto replied:
I'm using this map to verify if a potential CSI driver is installed on the node and report the appropriate error.

ravisantoshgudimetla commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -134,33 +137,36 @@ func (c *CSIMaxVolumeLimitChecker) filterAttachableVolumes(
 			continue
 		}
 
-		driverName, volumeHandle := c.getCSIDriver(pvc)
```
Please change this to V(5) for sake of consistency.

bertinatto replied:
Changed

ravisantoshgudimetla commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -134,33 +137,36 @@ func (c *CSIMaxVolumeLimitChecker) filterAttachableVolumes(
 
-		driverName, volumeHandle := c.getCSIDriver(pvc)
+		csiNode := nodeInfo.CSINode()
+		driverName, volumeHandle := c.getCSIDriverInfo(csiNode, pvc)
```
What if we return "", ""  but at https://github.com/kubernetes/kubernetes/pull/77595/commits/396b38f9bd2a6bd494112d4451134976bb598afc#diff-950de94621e8191fe9ca3368eef89318R208, shouldn't we log here?

bertinatto replied:
Added

msau42 commented:
I ran the secrets and secretsWithCSINode benchmarks with and without this pr on my local desktop with 12 cores.  And saw no signficant regression (within +/- 3%).  I also ran it with and without optimizations, and saw no significant difference either.

ravisantoshgudimetla commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -150,6 +155,63 @@ func (sched *Scheduler) deleteNodeFromCache(obj interface{}) {
+func (sched *Scheduler) onCSINodeUpdate(oldObj, newObj interface{}) {
+	oldCSINode, ok := oldObj.(*storagev1beta1.CSINode)
+	if !ok {
+		klog.Errorf("cannot convert oldObj to *v1.Node: %v", oldObj)
```
typo: to storage object?

bertinatto replied:
Fixed

gnufied committed:
Add unit tests for CSI predicate

ravisantoshgudimetla reviewed (commented):
@bertinatto & other folks - Thanks a lot for working on this PR.

Storage folks, while the code changes look good to me and the performance numbers look promising, I think we're too close to the release to merge this PR. I think, it'd be better to have enough soak time for this change so that some scalability tests on kubemark and real cluster could be run and we can prove this change doesn't have a performance impact(I am not denying that whatever data we've proves that impact is negligible) but what I have noticed in the past is - while benchmarks show improvements scalability tests on both kubemark and real cluster showed decrease in scheduling latency or throughput or both. So, I think it's risky atm to get this PR merged.

gnufied commented:
@misterikkit @ravisantoshgudimetla thank for the review guys. As I understand - there is nothing wrong in PR and performance numbers looks good but since we are too close to the release time to have proper soak time and hence we should try and get this PR in for the next release.

I think that is fine and we will try and get this in 1.16. cc @bertinatto @davidz627

jsafrane commented:
I tested with today's version (morning) and numbers seem to be better:


Test | Before | After | Diff
-- | -- | -- | --
BenchmarkSchedulingCSIPVs | 61923461.6 | 69684354 | 12.53%
BenchmarkSchedulingInTreePVs | 76875485.9 | 78601283.5 | 2.24%
BenchmarkSchedulingMigratedInTreePVs | 77419368.5 | 79149593.9 | 2.23%
BenchmarkSchedulingPodAntiAffinity | 36619170.1 | 36661907.3 | 0.12%
BenchmarkSchedulingSecrets | 29228343.3 | 28981172.3 | -0.85%
BenchmarkSchedulingSecretsWithCSINodes | 28915431.4 | 30266347 | 4.67%

gnufied commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -51,60 +65,67 @@ func NewCSIMaxVolumeLimitPredicate(
+	for volumeLimitKey, vt := range newVolumeCount {
 		maxVolumeLimit, ok := nodeVolumeLimits[v1.ResourceName(volumeLimitKey)]
-		if ok {
+		if !ok && vt.volumeType == csiVolumeType {
```
@bertinatto @msau42 So I made this code so that, if we know that volume is really of CSI type then and only then we prevent pod scheduling. For other cases we do the regular stuff. Also added unit tests for bunch of scenarios from here.

bertinatto replied:
> if we know that volume is really of CSI type

We still can't tell if the volume is CSI unless the translation library knows it. 

So the predicate will stop scheduling pods with unbound volumes (i.e., when the SC has `volumeBindingMode: WaitForFirstConsumer`) on nodes where the driver is not installed, **if** the volume belongs to a group of CSI drivers (EBS, GCEPD, etc.) 

However, the behaviour will be different for all other CSI drivers: k8s will keep scheduling pods on nodes where the driver is not installed.

IMO we should have a consistent behaviour among all CSI drivers, and keep the current behaviour at least until we can reliable tell whether an unbound volume is CSI or not.

bertinatto replied:
An option to solve this would be to have the predicate:

1) Look at the SC provisioner name (if the volume is unbound)
2) Then check if there's a `CSIDriver` object with the same name

We'd have to cache another object, but that give us confidence that the volume is really CSI. I discussed this with @jsafrane and he mentioned that `CSIDriver` object is currently optional, so we'd have to make it non-optional as well.

gnufied replied:
Yeah - we can use `CSIDriver` object but since it will require informer etc, that can be added as a different PR.

>However, the behaviour will be different for all other CSI drivers: k8s will keep scheduling pods on nodes where the driver is not installed.

This applies only when a PVC is being bound for the first time, in all other cases when PVC is already bound we will correctly stop pods from scheduling if driver is not installed. 

> IMO we should have a consistent behaviour among all CSI drivers, and keep the current behaviour at least until we can reliable tell whether an unbound volume is CSI or not.

Even if we use `CSIDriver` object the proposed feature (of preventing pod scheduling) will only work when `CSIDriver` object is installed. So it will always be a opt-in. 

We have two options:

1. Remove new behaviour entirely, because not counting PVCs if they are unbound is not an option. It will result in regression.
2. Keep counting unbound PVCs and only prevent pod scheduling (for no driver case) when we know for sure that PVC is of CSI type. 

We can extend#2 with `CSIDriver` check in future PR. I prefer option#2 rather than dropping 
"new behaviour" altogether. I am not sure about consistency argument - scheduler does its best to prevent scheduling whenever it can correctly determine PVC is of `CSI` type.

bertinatto replied:
> I am not sure about consistency argument - scheduler does its best to prevent scheduling whenever it can correctly determine PVC is of CSI type.

I used the consistency argument because given the exact same conditions, k8s will act differently depending on the CSI driver.

Another thing I was thinking is that a CSI driver author will have little to no power to change that, unless their driver is a target to an in-tree migration. The `CSIDriver` solution could be helpful in these cases.

IDK if my vote counts, but I'd keep this PR only focused on the volume limits and handle the "new behaviour" in another PR, where we can implement the `CSIDriver` look up only. 

That way, we'll be able to document that in order to prevent pods (with unbound PVCs) from scheduling in nodes where the driver is not installed, driver authors should make sure that a `CSIDriver` object is properly created for their driver.

jsafrane replied:
> IDK if my vote counts, but I'd keep this PR only focused on the volume limits and handle the "new behaviour" in another PR, where we can implement the CSIDriver look up only.

I vote for this too. Not sure if we need an extra predicate (because it would loop through all volumes in a pod and dereference PVCs again, as this predicate does), but it should be definitely a separate PR.

gnufied replied:
okay so just so I understand this. Even if we know volume being considered is of CSI type because it has PV source present, we are going to require `CSIDriver` object to prevent pod scheduling?  That is to ensure we use same thing as source of truth?

msau42 replied:
For PVC bound case, we can still use PV.CSISource.  For PVC unbound case, we can use CSIDriver object.  I'm fine with this behavior being opt-in.  Scheduling based on CSI driver node installation is a new feature and so far I'm not aware of any plugins that are going to be installed on only a subset of nodes.

msau42 replied:
I don't have strong opinions on whether or not it should be a separate PR or not.  Mostly depends on your bandwidth/confidence in implementing this by early 1.16 so we can get enough scalability soak time.  It would help make this already big PR easier to review potentially.

gnufied replied:
Alright I will leave this with @bertinatto . If you want to remove new behavior of preventing pod scheduling if driver is not installed on the node then feel free to delete all the code and tests related to it.

We will try and cover this in a different PR.

soggiest commented:
/milestone v1.16

bertinatto commented:
/test pull-kubernetes-kubemark-e2e-gce-big

gnufied commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -51,55 +54,52 @@ func NewCSIMaxVolumeLimitPredicate(
+	if len(nodeVolumeLimits) == 0 {
+		return true, nil, nil
+	}
+
```
Now that we don't care about preventing pod scheduling for volumes where driver is not installed, it might be cheaper to move this *after* `newVolumes == 0` check and return early. Evaluating all pods on a node is def. going to be costlier.

bertinatto replied:
Moved up

bertinatto commented:
/test pull-kubernetes-e2e-gce-csi-serial

bertinatto commented:
@ravisantoshgudimetla, @bsalamat, @misterikkit: now that the merge window is open, could you take another look? We're looking into merging this early in the cycle to have enough testing time.

bertinatto commented:
/retest

bertinatto commented:
/retest

bertinatto commented:
/retest

msau42 commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -51,51 +54,47 @@ func NewCSIMaxVolumeLimitPredicate(
-			delete(newVolumes, volumeName)
+	for volumeUniqueName, volumeLimitKey := range attachedVolumes {
+		if _, ok := newVolumes[volumeUniqueName]; ok {
+			delete(newVolumes, volumeUniqueName)
```
Can you add a comment here to say don't count multiple pods sharing the same volume twice.

msau42 commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -130,73 +129,102 @@ func (c *CSIMaxVolumeLimitChecker) filterAttachableVolumes(
+			return "", ""
+		}
+
+		if csiPV.Spec.PersistentVolumeSource.CSI == nil {
```
nil checks for each field that's dereferenced?

msau42 commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -130,73 +129,102 @@ func (c *CSIMaxVolumeLimitChecker) filterAttachableVolumes(
 	if pvName == "" {
 		klog.V(5).Infof("Persistent volume had no name for claim %s/%s", namespace, pvcName)
-		return c.getDriverNameFromSC(pvc)
+		return c.getCSIDriverInfoFromSC(pvc)
```
Do we need to handle migration here too to return migrated plugin name?

bertinatto replied:
That's a good point; migrated unbound PVCs wouldn't be counted...

Fixed and added a [test case](https://github.com/kubernetes/kubernetes/pull/77595/commits/745219bc1a406a76a16a85030d202adb50d275cb#diff-d1bf38bbaa6864388e575fcdcda652d4R335) to cover that.

msau42 commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -442,6 +442,8 @@ func ClusterRoles() []rbacv1.ClusterRole {
 				rbacv1helpers.NewRule("create").Groups(authenticationGroup).Resources("tokenreviews").RuleOrDie(),
 				rbacv1helpers.NewRule("create").Groups(authorizationGroup).Resources("subjectaccessreviews").RuleOrDie(),
+				// needed for checking volume limits on node
+				rbacv1helpers.NewRule(Read...).Groups(storageGroup).Resources("csinodes").RuleOrDie(),
```
This should be added under AttachLimits feature gate

msau42 commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate_test.go:
```diff
@@ -156,112 +213,249 @@ func TestCSIVolumeCountPredicate(t *testing.T) {
+			fits:             false,
+			migrationEnabled: true,
+			limitSource:      "csinode",
+			test:             "should count in-tree and csi volumes if migration is enabled (when scheduling csi volumes)",
```
the description is the same as the previous test case. can you clarify the difference?

msau42 commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate_test.go:
```diff
@@ -156,112 +213,249 @@ func TestCSIVolumeCountPredicate(t *testing.T) {
+			fits:             true,
+			migrationEnabled: false,
+			limitSource:      "csinode",
+			test:             "should not count in-tree and count csi volumes if migration is disabled (when scheduling csi volumes)",
```
same thing duplicate description

msau42 commented:
@jsafrane can you summarize your benchmark results again? For tests with CSI volumes and migration on/off, anything we need to investigate further?

bertinatto commented:
/retest

jsafrane commented:
Fresh numbers from today's morning:

| Test | Before (ns/op) | After (ns/op) | Increase
-- | -- | -- | --
BenchmarkSchedulingCSIPVs | 58977913.6 | 64527353.9 | 9.41%
BenchmarkSchedulingInTreePVs | 74486511.8 | 75411026.2 | 1.24%
BenchmarkSchedulingMigratedInTreePVs | 74846972.9 | 74396914.4 | -0.60%
BenchmarkSchedulingPodAntiAffinity | 36280202.1 | 36486112.5 | 0.57%
BenchmarkSchedulingSecrets | 29108418 | 29609940.5 | 1.72%
BenchmarkSchedulingSecretsWithCSINodes | 29691660.5 | 30196475 | 1.70%

Only scheduling of pods with CSI volumes is ~9% slower. The rest (incl. pod with Secrets and in-tree volumes) is negligible.

msau42 commented:
This LGTM from me. Will let sig-scheduling do final lgtm

jsafrane commented:
Just a note to the benchmark numbers: they were taken on 4 core machine. With 8 cores, there is no performance regression. Scheduler seems to run (some?) predicates in parallel and the updated CSIMaxVolumeLimitChecker predicate is still faster than the slowest already existing predicate.

ravisantoshgudimetla commented:
> With 8 cores, there is no performance regression. Scheduler seems to run (some?) predicates 

Yes, scheduler runs predicates in parallel. You mean on 4 core machine, you noticed performance regression of 9% whereas on 8 core there is no performance regression. That sounds reasonable. LGTM from my side

/lgtm

@bsalamat - Do you mind doing a final review since you were involved in the design?

bsalamat commented on pkg/scheduler/algorithm/predicates/csi_volume_predicate.go:
```diff
@@ -51,51 +54,47 @@ func NewCSIMaxVolumeLimitPredicate(
-	// if node does not have volume limits this predicate should exit
-	if len(nodeVolumeLimits) == 0 {
+	// If the new pod doesn't have any volume attached to it, the predicate will always be true
+	if len(pod.Spec.Volumes) == 0 {
```
Could you please check this condition before checking the feature gate? It appears that checking the feature gate is not very fast. So, it is better to check volumes length before that. Alternatively, you can combine the two if statements:

```
if len(pod.Spec.Volumes) == 0 || !utilfeature.DefaultFeatureGate.Enabled... {
    return ...
}
```

bertinatto replied:
Done

bsalamat reviewed (commented):
The changes to the scheduler look good to me. Only a couple of minor comments.

bertinatto commented on pkg/apis/storage/types.go:
```diff
@@ -355,6 +355,20 @@ type CSINodeDriver struct {
+type VolumeNodeResources struct {
+	// Maximum number of unique volumes managed by the CSI driver that can be used on a node.
+	// A volume that is both attached and mounted on a node is considered to be used once, not twice.
+	// The same rule applies for an unique volume that is shared among multiple pods on the same node.
```
Fixed

bertinatto commented:
/hold cancel

bertinatto commented:
/test pull-kubernetes-kubemark-e2e-gce-big

bertinatto commented:
/retest

davidz627 commented:
@bsalamat I believe your comments are addressed and tests are passing
👏 for everyone's massive effort on this!

msau42 commented:
/priority important-soon

bsalamat reviewed (commented):
/lgtm
/approve

Thanks, @bertinatto!

tedyu commented on pkg/registry/storage/csinode/strategy.go:
```diff
@@ -62,8 +70,33 @@ func (csiNodeStrategy) AllowCreateOnUpdate() bool {
+	newCSINode := obj.(*storage.CSINode)
+	oldCSINode := old.(*storage.CSINode)
+
+	inUse := getAllocatablesInUse(oldCSINode)
```
It seems this call should be moved inside the feature gate check.

tedyu commented on pkg/scheduler/internal/cache/cache.go:
```diff
@@ -569,6 +570,47 @@ func (cache *schedulerCache) RemoveNode(node *v1.Node) error {
+	if !ok {
+		return fmt.Errorf("node %v is not found", csiNode.Name)
+	}
+	n.info.SetCSINode(nil)
```
If n, the NodeInfoListItem, was created on line 579, shouldn't cache.removeNodeInfoFromList be called ?

tedyu commented on pkg/scheduler/nodeinfo/node_info.go:
```diff
@@ -471,11 +482,24 @@ func (n *NodeInfo) Clone() *NodeInfo {
+			d := n.csiNode.Spec.Drivers[i]
+			if d.Allocatable != nil && d.Allocatable.Count != nil {
+				// TODO: drop GetCSIAttachLimitKey once we don't get values from Node object
+				k := v1.ResourceName(volumeutil.GetCSIAttachLimitKey(d.Name))
```
Can we establish map from ResourceName to int64 once so that we don't perform the calculation here every time ?

ProgramerGu replied:
What if we may need the ResourceName other place?

tedyu commented:
@jsafrane @msau42 
what do you think of the proposal of caching csiVolumeLimits inside NodeInfo ?

thanks

msau42 commented:
@tedyu the number of different drivers installed on a node are not expected to be very high. I'm not sure if we will see much performance benefit from this.  You can try running the benchmarks that @jsafrane developed to see.  https://github.com/kubernetes/kubernetes/pull/77397
